subsys/net/lib/rest_client/src/rest_client.c

```diff
@@ -0,0 +1,360 @@
+	} else {
+		cache = TLS_SESSION_CACHE_DISABLED;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_SESSION_CACHE, &cache, sizeof(cache));
+	if (err) {
+		LOG_ERR("Unable to set session cache, errno %d", errno);
+		return err;
+	}
+
+	if (tls_hostname) {
+		err = setsockopt(fd, SOL_TLS, TLS_HOSTNAME, tls_hostname, strlen(tls_hostname));
+		if (err) {
+			LOG_ERR("Failed to setup TLS hostname, error: %d", errno);
+			return err;
+		}
+	}
+	return 0;
+}
+
+static int rest_client_sckt_timeouts_set(int fd)
+{
+	int err;
+	struct timeval timeout = { 0 };
+
+	if (CONFIG_REST_CLIENT_SCKT_SEND_TIMEOUT > -1) {
+		/* Send TO also affects TCP connect */
+		timeout.tv_sec = CONFIG_REST_CLIENT_SCKT_SEND_TIMEOUT;
+		err = setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &timeout, sizeof(timeout));
+		if (err) {
+			LOG_ERR("Failed to set socket send timeout, error: %d", errno);
+			return err;
+		}
+	}
+
+	if (CONFIG_REST_CLIENT_SCKT_RECV_TIMEOUT > -1) {
+		timeout.tv_sec = CONFIG_REST_CLIENT_SCKT_RECV_TIMEOUT;
+		err = setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+		if (err) {
+			LOG_ERR("Failed to set socket recv timeout, error: %d", errno);
+			return err;
+		}
+	}
+	return 0;
+}
+
+static int rest_client_sckt_connect(int *const fd, const char *const hostname,
+				     const uint16_t port_num,
+				     const sec_tag_t sec_tag,
+				     int tls_peer_verify)
+{
+	int ret;
+	struct addrinfo *addr_info;
+	char peer_addr[INET_ADDRSTRLEN];
+	const char *const connect_addr = hostname;
+
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+		.ai_next = NULL,
+	};
+	int proto = 0;
+
+	/* Make sure fd is always initialized when this function is called */
+	*fd = -1;
+
+	LOG_DBG("Doing getaddrinfo() with connect addr %s", log_strdup(connect_addr));
+
+	ret = getaddrinfo(connect_addr, NULL, &hints, &addr_info);
+	if (ret) {
+		LOG_ERR("getaddrinfo() failed, error: %d", ret);
+		return -EFAULT;
+	}
+
+	inet_ntop(AF_INET, &net_sin(addr_info->ai_addr)->sin_addr, peer_addr, INET_ADDRSTRLEN);
+	LOG_DBG("getaddrinfo() %s", log_strdup(peer_addr));
+
+	((struct sockaddr_in *)addr_info->ai_addr)->sin_port = htons(port_num);
+
+	proto = (sec_tag == REST_CLIENT_NO_SEC) ? IPPROTO_TCP : IPPROTO_TLS_1_2;
+	*fd = socket(AF_INET, SOCK_STREAM, proto);
+	if (*fd == -1) {
+		LOG_ERR("Failed to open socket, error: %d", errno);
+		ret = -ENOTCONN;
+		goto clean_up;
+	}
+
+	if (sec_tag >= 0) {
+		ret = rest_client_sckt_tls_setup(*fd, hostname, sec_tag, tls_peer_verify);
+		if (ret) {
+			ret = -EACCES;
+			goto clean_up;
+		}
+	}
+
+	ret = rest_client_sckt_timeouts_set(*fd);
+	if (ret) {
+		LOG_ERR("Failed to set socket timeouts, error: %d", errno);
+		ret = -EINVAL;
+		goto clean_up;
+	}
+
+	LOG_DBG("Connecting to %s", log_strdup(connect_addr));
+
+	ret = connect(*fd, addr_info->ai_addr, sizeof(struct sockaddr_in));
+	if (ret) {
+		LOG_ERR("Failed to connect socket, error: %d", errno);
+		ret = -ECONNREFUSED;
+		goto clean_up;
+	}
+
+	return ret;
```

Comment:
The success case should probably also flow through to `clean_up`. Now `freeaddrinfo()` does not get called if `connect()` succeeds.